src/SourceBuild/content/eng/tools/tasks/Microsoft.DotNet.SourceBuild.Tasks.XPlat/UpdateNuGetConfigPackageSourcesMappings.cs

```diff
@@ -34,48 +40,113 @@ public class UpdateNuGetConfigPackageSourcesMappings : Task
+                document.Root.Add(pkgSrcMappingElement);
             }
 
-            // Union all package sources to get the distinct list.  These will get added to the source-build sources.
-            string[] packagePatterns = pkgSrcMappingElement.Descendants()
-                .Where(e => e.Name == "packageSource")
-                .SelectMany(e => e.Descendants().Where(e => e.Name == "package"))
-                .Select(e => e.Attribute("pattern").Value)
-                .Distinct()
-                .ToArray();
+            DiscoverPackagesFromAllSourceBuildSources(pkgSourcesElement);
 
-            if (!BuildWithOnlineFeeds)
+            // Discover all SBRP packages if source-build-reference-package-cache source is present in NuGet.config
+            XElement sbrpCacheSourceElement = pkgSourcesElement.Descendants().FirstOrDefault(e => e.Name == "add" && e.Attribute("key").Value == SbrpCacheSourceName);
+            if (sbrpCacheSourceElement != null)
             {
-                // When building offline remove all packageSourceMappings.
-                pkgSrcMappingElement?.ReplaceNodes(new XElement("clear"));
+                DiscoverPackagesFromSbrpCacheSource();
             }
 
+            // If building online, enumerate any existing package source mappings and filter
+            // to remove packages that are present in any local source-build source
+            if (BuildWithOnlineFeeds && pkgSrcMappingElement != null)
+            {
+                GetExistingFilteredSourceMappings(pkgSrcMappingElement);
+            }
+
+            // Remove all packageSourceMappings
+            pkgSrcMappingElement.ReplaceNodes(new XElement("clear"));
+
             XElement pkgSrcMappingClearElement = pkgSrcMappingElement.Descendants().FirstOrDefault(e => e.Name == "clear");
-            if (pkgSrcMappingClearElement == null)
+
+            // Add package source mappings for local package sources
+            foreach (string packageSource in allSourcesPackages.Keys)
             {
-                pkgSrcMappingClearElement = new XElement("clear");
-                pkgSrcMappingElement.AddFirst(pkgSrcMappingClearElement);
+                // Skip sources with zero package patterns
+                if (allSourcesPackages[packageSource] != null)
+                {
+                    pkgSrcMappingClearElement.AddAfterSelf(GetPackageMappingsElementForSource(packageSource));
+                }
             }
 
-            foreach (string packageSource in SourceBuildSources)
+            // When building online add the filtered mappings from original online sources.
+            // If there are none, add default mappings for all online sources.
+            if (BuildWithOnlineFeeds)
             {
-                XElement pkgSrc = new XElement("packageSource", new XAttribute("key", packageSource));
-                foreach (string packagePattern in packagePatterns)
+                if (oldSourceMappingPatterns.Count > 0)
```

Comment:
Nit, this if condition seems unnecessary.  It is adding unnecessary complexity and nesting, the foreach will no-op with an empty collection.